gather.go

```diff
@@ -644,77 +648,207 @@ func (a *Agent) gatherCandidatesRelay(ctx context.Context, urls []*stun.URI) { /
+	// just in case we feed the relayAllocation address through a resolution step.
+	relayAddr := relayAllocation.Addr()
+	relayTCPAddr, err := net.ResolveTCPAddr("tcp", relayAddr.String())
+	if err != nil {
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to resolve relay address %s %s", relayAddr, err)
+		return
+	}
+
+	relayConfig := CandidateRelayConfig{
+		Network:         network,
+		Component:       ComponentRTP,
+		Address:         relayTCPAddr.IP.String(),
+		Port:            relayTCPAddr.Port,
+		RelAddr:         relatedAddr,
+		RelPort:         relatedPort,
+		TURNControlConn: client,
+		// Dan: We force TCPTypePassive here by choice. Maybe SimultaneousOpen is valid as well? We
+		// certainly expect to only succeed via the other peer dialing out to the relay. But if we
+		// have firewall/nat limitations on the TURN server, simultaneous open may have an
+		// advantage? Can certainly experiment here at a later date.
+		TCPType:       TCPTypePassive,
+		RelayProtocol: tcp,
+		OnClose: func() error {
+			relayAllocation.Close()
+			client.Close()
+			locConn.Close()
+			return nil
+		},
+	}
+
+	// If we do not succeed in adding this candidate, we must clean up
+	relayConnClose := func() {
+		if relayConErr := relayAllocation.Close(); relayConErr != nil {
+			a.log.Warnf("Failed to close relay %v", relayConErr)
+		}
+	}
+
+	candidate, err := NewCandidateRelay(&relayConfig)
+	if err != nil {
+		relayConnClose()
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to create relay candidate: %s %s: %v", network, relayAddr.String(), err)
+		return
+	}
+
+	if err := a.addCandidate(ctx, candidate, dataConn); err != nil {
+		relayConnClose()
+		if closeErr := candidate.close(); closeErr != nil {
+			a.log.Warnf("Failed to close candidate: %v", closeErr)
+		}
+		a.log.Warnf("Failed to append to localCandidates and run onCandidateHdlr: %v", err)
+	}
+}
+
+func (a *Agent) createUDPRelayCandidate(
+	ctx context.Context, network, turnServerAddr string, locConn net.PacketConn, url stun.URI,
+	relatedAddr string, relatedPort int, relayProtocol string) {
+
+	client, err := turn.NewClient(&turn.ClientConfig{
+		TURNServerAddr: turnServerAddr,
+		Conn:           locConn,
+		Username:       url.Username,
+		Password:       url.Password,
+		LoggerFactory:  a.loggerFactory,
+		Net:            a.net,
+	})
+	if err != nil {
+		closeConnAndLog(locConn, a.log, "failed to create new TURN client %s %s", turnServerAddr, err)
+		return
+	}
+
+	if err = client.Listen(); err != nil {
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to listen on TURN client %s %s", turnServerAddr, err)
+		return
+	}
+
+	relayConn, err := client.Allocate()
+	if err != nil {
+		client.Close()
+		closeConnAndLog(locConn, a.log, "failed to allocate on TURN client %s %s", turnServerAddr, err)
+		return
+	}
+
+	relayAddr := relayConn.LocalAddr().(*net.UDPAddr) //nolint:forcetypeassert
```

Comment:
No resolution as you had above?

Reply:
This is the existing code. I believe I left it alone.

The reason for the resolution above was to turn the udpAddr into a tcpAddr. Such that when the `AllocateTCP` method is called, the code can safely assume it has "tcp addresses". I think. I don't believe there's an actual dns resolution here.